File: backend/app/models/risk.py
from pydantic import BaseModel
from typing import Optional, Literal, List
from datetime import date


class UncertaintyRange(BaseModel):
    activity_id: str
    optimistic_days: int
    most_likely_days: int
    pessimistic_days: int
    distribution: Literal["pert", "triangular", "uniform"] = "pert"


class SimulationConfig(BaseModel):
    project_id: str
    iterations: int = 10000
    ranges: List[UncertaintyRange]


class DistributionBucket(BaseModel):
    date: date
    count: int
    cumulative_probability: float


class SensitivityItem(BaseModel):
    activity_id: str
    activity_name: str
    correlation: float


class SimulationResult(BaseModel):
    p50_date: date
    p80_date: date
    p95_date: date
    deterministic_date: date
    completion_distribution: List[DistributionBucket]
    sensitivity: List[SensitivityItem]
    total_iterations: int
    execution_time_ms: int

File: frontend/src/types/risk.ts
export interface UncertaintyRange {
  activity_id: string;
  optimistic_days: number;
  most_likely_days: number;
  pessimistic_days: number;
  distribution: 'pert' | 'triangular' | 'uniform';
}

export interface SimulationConfig {
  project_id: string;
  iterations: number;
  ranges: UncertaintyRange[];
}

export interface DistributionBucket {
  date: string;
  count: number;
  cumulative_probability: number;
}

export interface SensitivityItem {
  activity_id: string;
  activity_name: string;
  correlation: number;
}

export interface SimulationResult {
  p50_date: string;
  p80_date: string;
  p95_date: string;
  deterministic_date: string;
  completion_distribution: DistributionBucket[];
  sensitivity: SensitivityItem[];
  total_iterations: number;
  execution_time_ms: number;
}
